### PR TITLE
Drop head

### PR DESCRIPTION
**Pull request by T-256**
I didn't understand why it is there!?
Perhaps to prevent annotations conflicts? then there is `from __future__ import annotations` for it.
This PR will reduce 5% of `Win32` package size.

## Test Plan
`Win32`: generated and ran imports successfully.
`WinRT`: not tested

### CODE REVIEW

**Comment by T-256** [comment]
> I didn't understand why it is there!?

Ok, got it now, it's for `press` function.
I'm going to change it as decorator.

File: win32more/__init__.py [hunk 1]
```diff
@@ -354,64 +385,17 @@ def wrapper(*args, **kwargs):
...
-        types = types[-1:] + types[:-1]
-        return functype(*types)
-
-    return press_functype_pointer
-
 
 def cfunctype_pointer(prototype):
-    return commonfunctype_pointer(prototype, CFUNCTYPE)
+    hints = get_type_hints(prototype)
+    types = list(hints.values())
```
Comment (T-256) [inline win32more/__init__.py:391]:
```suggestion
    types = list(get_type_hints(prototype).values())

```

File: win32more/__init__.py [hunk 2]
```diff
@@ -354,64 +385,17 @@ def wrapper(*args, **kwargs):
...
+    types = types[-1:] + types[:-1]
+    return CFUNCTYPE(*types)
 
 
 def winfunctype_pointer(prototype):
-    return commonfunctype_pointer(prototype, WINFUNCTYPE)
+    hints = get_type_hints(prototype)
+    types = list(hints.values())
```
Comment (T-256) [inline win32more/__init__.py:398]:
```suggestion
    types = list(get_type_hints(prototype).values())

```

File: win32more/__init__.py [hunk 3]
```diff
@@ -354,64 +385,17 @@ def wrapper(*args, **kwargs):
...
-        types = types[-1:] + types[:-1]
-        return functype(*types)
-
-    return press_functype_pointer
-
 
 def cfunctype_pointer(prototype):
-    return commonfunctype_pointer(prototype, CFUNCTYPE)
+    types = list(get_type_hints(prototype).values())
+
```
Comment (T-256) [inline win32more/__init__.py:391]:
```suggestion
```

File: win32more/__init__.py [hunk 4]
```diff
@@ -354,64 +385,17 @@ def wrapper(*args, **kwargs):
...
+    types = types[-1:] + types[:-1]
+    return CFUNCTYPE(*types)
 
 
 def winfunctype_pointer(prototype):
-    return commonfunctype_pointer(prototype, WINFUNCTYPE)
+    types = list(get_type_hints(prototype).values())
```
Comment (T-256) [inline win32more/__init__.py:398]:
```suggestion
```

**Comment by T-256** [comment]
TODO: Don't re-import self-path in modules.

**Comment by T-256** [comment]
for fixing this error:
```
AttributeError: module 'win32more.Windows.Win32.Graphics' has no attribute 'Gdi'
```
I suggest handle such cases and register them in metaclass, then retry to solve them at end of each module.
Also can consider lazy loading until user explicitly use structs.

**Comment by ynkdir** [comment]
__getattr__ and make_head are used for circular dependency.

**Comment by T-256** [comment]
> **getattr** and make_head are used for circular dependency.

Yes, I got it,
Current approach I used is metaclassing lazy loading rather than module level lazy loading.

I think besides file sizes, it improves both cpu & memory usages.

File: win32more/__init__.py [hunk 5]
```diff
@@ -174,6 +226,27 @@ def easycast(obj, type_):
...
+
+def get_hints(struct):
+    hints = {}
+    for hint, type_ in get_type_hints(struct).items():
+        type_ = _patch_char_p(type_)
+        if not hasattr(type_, '__done_ctypes__'):
+            if issubclass(type_, (EasyCastStructure, EasyCastUnion)):
+                EasyCastMeta.commit(type_)
+            elif issubclass(type_, ComPtr):
+                ComPtr.commit(type_)
```
Comment (T-256) [inline win32more/__init__.py:245]:
```suggestion
        if not hasattr(type_, '__done_ctypes__'):
            if isinstance(type_, BaseFuncType):
                if not hasattr(type_, '__done_ctypes__'):
                    type_.__done_ctypes__ = True
                    BaseFuncType.commit(type_)
                type_ = type_._fn
            elif issubclass(type_, (EasyCastStructure, EasyCastUnion)):
                EasyCastMeta.commit(type_)
            elif issubclass(type_, ComPtr):
                ComPtr.commit(type_)
```

File: win32more/__init__.py [hunk 6]
```diff
@@ -67,8 +68,29 @@
...
+            cls.__done_ctypes__ = True
+            ComPtrMeta.commit(cls)
+        return super().__call__(*args, **kwargs)
+
+    @classmethod
+    def commit(cls, struct):
+        struct._hints_ = get_hints(struct)
+
+        if struct._hints_["extends"] is None:
+            return
```
Comment (T-256) [inline win32more/__init__.py:83]:
```suggestion
        if struct.__annotations__["extends"] == 'None':
            return

        struct._hints_ = get_hints(struct)
```

**Comment by T-256** [comment]
POINT should initialize once to `_fields_` apply on it.
I think these should work:
```py
POINT()
assert sizeof(POINT) == 8
```
OR
```py
assert sizeof(POINT()) == 8
```

AFAIK only `sizeof` has this problem and I also encountered this too, then I used custom `sizeof`:
```py
def sizeof(obj, *args, *kwargs):
    if (
        isinstance(obj, type)
        and issubclass(obj, (EasyCastStructure, EasyCastUnion, ComPtr))
        and not hasattr(obj, '__done_ctypes__')
    ): obj()

    return ctypes.sizeof(obj, *args, **kwargs)
```

You can accept these for now or wait for https://github.com/T-256/py-win32more/pull/1 to be complemented.
The Idea is using `__getattribute__` at end of modules that that triggers metaclass `commit` method (whats `__call__` doing now).

**Comment by T-256** [comment]
closing in favor of #17 which doesn't have `sizeof` problem (uses module's `__gettattr__` instead of metaclass's `__call__`)

**Pull request closed without merging**